### PR TITLE
feature: search

### PR DESCRIPTION
**Pull request by ireznice**

### CODE REVIEW

File: lib/final-api/v1/http/ddtf_build.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ class DDTF_Build
           'failed' => 'Finished',
           'canceled' => 'Stopped',
           'errored' => 'Aborted',
-          '' => 'Unknown'
+          '' => 'Unbeknownst'
```
Comment (petrSchreiber) [inline lib/final-api/v1/http/ddtf_build.rb:21]: Extra points for style

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        count_wanted = builds.keep_if {|i| i.state == 'passed'}.size
+        expect(results.length).to eq(count_wanted)
+        results.each do |build|
+          expect(build.state).to eq('passed')
+        end
+      end
+    end
+  end
+
+  describe '::patse_query' do
```
Comment (ondrej-hosak) [inline spec/features/lib/final-api/model_extensions_spec.rb]: typo?

File: lib/final-api/v1/http/ddtf_build.rb [hunk 3]
```diff
@@ -136,8 +136,9 @@ def execution_logs
         private
 
         def product
-           tsd = build.config[:tsdContent]
-           tsd[:product] if tsd
+          tsd = build.config[:tsdContent]
+          # HACK: avoiding the situation when tsd is a string accidentally
+          tsd[:product] if tsd && tsd.is_a?(Hash)
```
Comment (ondrej-hosak) [inline lib/final-api/v1/http/ddtf_build.rb]: this hack if not needed

File: lib/final-api/model_extensions.rb [hunk 4]
```diff
@@ -83,6 +48,68 @@ def sanitize
...
+      if exact_match
+        out << new if old.downcase == query.downcase
+      else
+        out << new if old.downcase.include? query.downcase
+      end
+    end.compact
+  end
+
+  def self.retrieve_filter(expr)
+    exact_match = (expr[1] == '=')
```
Comment (petrSchreiber) [inline lib/final-api/model_extensions.rb]: expr[0] is used multiple times, which means dereferencing each time, what about:
field = expr[0]
operator = expr[1]
value = expr[2]

File: lib/final-api/model_extensions.rb [hunk 5]
```diff
@@ -3,68 +3,33 @@
 class Build
   include FinalAPI::DDTF
 
+  class InvalidQueryError < StandardError
+  end
+
   # Represents metadat for query language used by frondend
```
Comment (petrSchreiber) [inline lib/final-api/model_extensions.rb]: metadat -> metadata

File: lib/final-api/model_extensions.rb [hunk 6]
```diff
@@ -83,6 +48,68 @@ def sanitize
...
+      User.where(name: query).each_with_object([]) {|u,out| out << u.id }
+    else
+      User.where("name ILIKE :expr", expr: "%#{query}%").each_with_object([]) {|u,out| out << u.id }
+    end
+  end
+
+  # maps fragment of old state given to travis states
+  def self.determine_states(query, exact_match = false)
+    states_map = FinalAPI::V1::Http::DDTF_Build::BUILD_STATE2API_V1STATUS
+    states_map.reject { |k,v| k == '' }.each_with_object([]) do |(new, old), out|
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb]: please rename new variable

File: lib/final-api/model_extensions.rb [hunk 7]
```diff
@@ -3,68 +3,33 @@
...
     ['id']                    => 'id',  #where(id OP id)
     ['nam', 'name']           => 'name',
-    ['sta', 'startedBy']      => 'owner_id',  # where('owner_id IN (?)', User.where('name ILIKE ?', '%KEY%'))
-    ['sto', 'stoppedBy']      => 'stopped_by_id',
+    ['sta', 'startedby']      => 'owner_id',  # where('owner_id IN (?)', User.where('name ILIKE ?', '%KEY%'))
+    ['sto', 'stoppedby']      => 'stopped_by_id',
     ['sts', 'stat', 'status', 'state'] => 'state',
     ['bui', 'build']          => 'build_info',
-    ['buildId', 'protonId']  => 'protonId'
+    ['buildid', 'protonid']  => 'proton_id'
```
Comment (ondrej-hosak) [inline lib/final-api/model_extensions.rb]: remove buildid, it will be no longer supported

File: lib/final-api/model_extensions.rb [hunk 8]
```diff
@@ -83,6 +48,68 @@ def sanitize
...
+  #
+  # For example:
+  #   parse_query('nam:"foo bar baz" bui =qux id : 1')
+  #     => [ ['nam', ':', 'foo bar baz'], ['bui', '=', 'qux'], ['id', ':', '1']]
+  def self.parse_query(query)
+    array = query.scan(/([^\s]*)\s*([:=])\s*("[^"]*"|\S*)/)
+    return [['name' , ':', query]] if array.length == 0
+    wrong_keys = []
+    result = array.map do |item|
+      query_key = item[0].downcase
```
Comment (petrSchreiber) [inline lib/final-api/model_extensions.rb:59]: I would suggest to extract item [1] and [2] to descriptive variables too

File: lib/final-api/endpoint/ddtf.rb [hunk 9]
```diff
@@ -58,8 +58,7 @@ def registered(app)
           app.get '/ddtf/tests' do
             limit = params[:limit] ? params[:limit].to_i : 20
             offset = params[:offset] ? params[:offset].to_i : 0
-            builds = Build.order(Build.arel_table['created_at'].desc).limit(limit).offset(offset)
-            builds = builds.ddtf_search(params[:q])
+            builds = Build.search(params[:q], limit, offset)
 
             # HACK: workaround make builds valid, could be removed later, when DB
             # will be valid
```
Comment (ondrej-hosak) [inline lib/final-api/endpoint/ddtf.rb]: please remove sanitize call, its no longer needed

File: lib/final-api/model_extensions.rb (hunk 6 above)
Comment (petrSchreiber) [inline lib/final-api/model_extensions.rb]: when does the k == '' happen?

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+      context 'when id does not exist' do
+        it 'returns an empty array' do
+          expect(Build.search("id = \"99999999999999\" ", 5, 0).length).to eq(0)
+        end
+      end
+
+      it 'filters by startedBy' do
+        builds << FactoryGirl.create(:build_customized, owner: user2)
+        filtered = Build.search("sta = quux.qux", 5, 0)
```
Comment (ireznice) [inline spec/features/lib/final-api/model_extensions_spec.rb]: user2.name

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+        expect(filtered.first.owner.name).to eq('quux.qux')
+        expect(filtered.length).to eq(1)
+      end
+
+      it 'filters by startedBy' do
+        builds << FactoryGirl.create(:build_customized, owner: user2)
+        filtered = Build.search("sta = franta.lopata", 500, 0)
```
Comment (ireznice) [inline spec/features/lib/final-api/model_extensions_spec.rb]: as above

File: spec/features/lib/final-api/model_extensions_spec.rb [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+
+      context 'when stoppedBy does not exist' do
+        it 'returns no items' do
+          builds << FactoryGirl.create(:build_customized, stopped_by: user1, owner: user2)
+          filtered = Build.search("sto= \"frant\" ", 500, 0)
+          expect(filtered.length).to eq(0)
+        end
+      end
+
+      it 'filters by state Configured' do
```
Comment (ireznice) [inline spec/features/lib/final-api/model_extensions_spec.rb]: does not filter by incomplete state name

File: lib/final-api/endpoint/ddtf.rb [hunk 13]
```diff
@@ -56,21 +56,23 @@ def registered(app)
...
+              limit = params[:limit] ? params[:limit].to_i : 20
+              offset = params[:offset] ? params[:offset].to_i : 0
+              builds = Build.search(params[:q], limit, offset)
 
-            FinalAPI::V1::Http::DDTF_Builds.new(builds, {}).data.to_json
+              FinalAPI::V1::Http::DDTF_Builds.new(builds, {}).data.to_json
+            rescue Build::InvalidQueryError => e
+              halt 404, {
+                status: 'error',
+                message: e.to_s
```
Comment (ondrej-hosak) [inline lib/final-api/endpoint/ddtf.rb]: return code 400 or find a suitable one. Hash needs to contain 'data' key.

**Pull request closed without merging**